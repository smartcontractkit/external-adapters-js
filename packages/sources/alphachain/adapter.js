const { Requester, Validator } = require('@chainlink/external-adapter')
const { util } = require('@chainlink/ea-bootstrap')

const customError = (data) => {
  if (data.status !== '200') return true
  return false
}

const customParams = {
  base: ['base', 'from', 'coin'],
  quote: ['quote', 'to', 'market'],
}

const execute = (input, callback) => {
  const validator = new Validator(input, customParams)
  if (validator.error) return callback(validator.error.statusCode, validator.errored)

  const jobRunID = validator.validated.id
  const url = 'https://alpha-chain2.p.rapidapi.com/data-query'
  const host = 'alpha-chain2.p.rapidapi.com'
  const headers = {
    'content-type': 'application/octet-stream',
    'x-rapidapi-host': host,
    'x-rapidapi-key': util.getRandomRequiredEnv('API_KEY'),
    useQueryString: true,
  }
  const base = validator.validated.data.base.toUpperCase()
  const quote = validator.validated.data.quote.toUpperCase()

  const params = {
    from_symbol: base,
    to_symbol: quote,
    chainlink_node: true,
  }

  const config = {
    url,
    params,
    headers,
  }

  Requester.request(config, customError)
    .then((response) => {
      response.data.result = Requester.validateResultNumber(response.data, ['result'])
      callback(response.status, Requester.success(jobRunID, response))
    })
    .catch((error) => callback(500, Requester.errored(jobRunID, error)))
}

module.exports.execute = execute
