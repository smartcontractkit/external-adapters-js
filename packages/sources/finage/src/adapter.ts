import { Execute } from '@chainlink/types'
import { Requester, Validator } from '@chainlink/external-adapter'
import { util } from '@chainlink/ea-bootstrap'

const customParams = {
  symbol: ['base', 'from', 'symbol'],
  to: false,
  endpoint: false,
}

const commonKeys: Record<string, string> = {
  FTSE: 'UK100',
  N225: 'JAP225',
}

export const execute: Execute = async (input) => {
  const validator = new Validator(input, customParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  const endpoint = validator.validated.data.endpoint || ''
  let url = `https://api.finage.co.uk/last/${endpoint}`
  const symbol = validator.validated.data.symbol.toUpperCase()
  const to = (validator.validated.data.to || '').toUpperCase()
  const currencies = (commonKeys[symbol] || symbol) + to
  const apikey = util.getRandomRequiredEnv('API_KEY')
  let params
  let responsePath

  switch (endpoint) {
    case 'stock': {
      url = `${url}/${symbol}`
      responsePath = ['bid']
      params = {
        apikey,
      }
      break
    }
    default: {
      responsePath = ['currencies', 0, 'value']
      params = {
        currencies,
        apikey,
      }
      break
    }
  }

  const config = {
    url,
    params,
  }

  const response = await Requester.request(config)
  response.data.result = Requester.validateResultNumber(response.data, responsePath)
  return Requester.success(jobRunID, response)
}
