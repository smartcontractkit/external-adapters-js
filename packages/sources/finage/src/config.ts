import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as BaseConfig } from '@chainlink/types'

export const NAME = 'FINAGE'

export const DEFAULT_BASE_URL = 'https://api.finage.co.uk'
export const DEFAULT_STOCK_WS_API_ENDPOINT = 'wss://e4s39ar3mr.finage.ws:7002'
export const DEFAULT_FOREX_WS_API_ENDPOINT = 'wss://w29hxx2ndd.finage.ws:8001'
export const DEFAULT_CRYPTO_WS_API_ENDPOINT = 'wss://72x8wsyx7t.finage.ws:6008'
export const DEFAULT_ENDPOINT = 'stock'

export const ENV_STOCK_WS_API_ENDPOINT = 'STOCK_WS_API_ENDPOINT'
export const ENV_FOREX_WS_API_ENDPOINT = 'FOREX_WS_API_ENDPOINT'
export const ENV_CRYPTO_WS_API_ENDPOINT = 'CRYPTO_WS_API_ENDPOINT'

export type Config = BaseConfig & {
  stockWsEndpoint: string
  forexWsEndpoint: string
  cryptoWsEndpoint: string
}

export const makeConfig = (prefix?: string): Config => {
  const config = Requester.getDefaultConfig(prefix, true)
  const stockWsUrl = util.getEnv(ENV_STOCK_WS_API_ENDPOINT) || DEFAULT_STOCK_WS_API_ENDPOINT
  const forexWsUrl = util.getEnv(ENV_FOREX_WS_API_ENDPOINT) || DEFAULT_FOREX_WS_API_ENDPOINT
  const cryptoWsUrl = util.getEnv(ENV_CRYPTO_WS_API_ENDPOINT) || DEFAULT_CRYPTO_WS_API_ENDPOINT
  const socketKey = util.getEnv('WS_SOCKET_KEY')
  return {
    ...config,
    api: {
      ...config.api,
      baseURL: config.api.baseURL || DEFAULT_BASE_URL,
    },
    stockWsEndpoint: `${stockWsUrl}?token=${socketKey}`,
    forexWsEndpoint: `${forexWsUrl}?token=${socketKey}`,
    cryptoWsEndpoint: `${cryptoWsUrl}?token=${socketKey}`,
    defaultEndpoint: DEFAULT_ENDPOINT,
  }
}
