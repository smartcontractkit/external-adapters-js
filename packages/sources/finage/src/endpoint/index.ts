import type { TInputParameters as StockInputParameters } from './stock'
import type { TInputParameters as EodInputParameters } from './eod'
import type { TInputParameters as ForexInputParameters } from './forex'
import type { TInputParameters as CryptoInputParameters } from './crypto'
import type { TInputParameters as CommoditiesInputParameters } from './commodities'

export type TInputParameters =
  | StockInputParameters
  | EodInputParameters
  | ForexInputParameters
  | CryptoInputParameters
  | CommoditiesInputParameters

export * as stock from './stock'
export * as eod from './eod'
export * as forex from './forex'
export * as crypto from './crypto'
export * as commodities from './commodities'
