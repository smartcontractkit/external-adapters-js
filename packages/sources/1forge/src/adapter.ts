import { Builder, Logger, Requester, Validator } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
  MakeWSHandler,
} from '@chainlink/ea-bootstrap'
import { DEFAULT_WS_API_ENDPOINT, makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (config?: DefaultConfig): MakeWSHandler => {
  const getSubscription = (symbol?: string, subscribe = true): string => {
    if (!symbol) return ''
    return subscribe ? `subscribe_to|${symbol}` : `unsubscribe_from|${symbol}`
  }
  const getSymbol = (input: AdapterRequest) => {
    const validator = new Validator(
      input,
      endpoints.quotes.inputParameters,
      {},
      { shouldThrowError: false },
    )
    if (validator.error) return
    if (Array.isArray(validator.validated.data.quote)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.quote} supplied as quote. Only non-array tickers can be used for WS`,
      )
      return
    }
    if (Array.isArray(validator.validated.data.base)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.base} supplied as base. Only non-array tickers can be used for WS`,
      )
      return
    }
    const symbol = validator.validated.data.base.toUpperCase()
    const convert = validator.validated.data.quote.toUpperCase()
    return `${symbol}/${convert}`
  }
  const parseResponse = (response: string) => {
    const [type, data] = response.split('|')
    return { type, data: JSON.parse(data) }
  }

  return () => {
    const defaultConfig = config || makeConfig()

    return {
      connection: {
        url: DEFAULT_WS_API_ENDPOINT,
      },
      subscribe: (input) => getSubscription(getSymbol(input)),
      unsubscribe: (input) => getSubscription(getSymbol(input), false),
      subsFromMessage: (message: any) => {
        if (!message.data || message.data.indexOf('update') === -1) return
        const { data } = parseResponse(message.data)
        return getSubscription(data.s)
      },
      isError: () => false,
      filter: (message: any) => message.data.indexOf('update') !== -1,
      toResponse: (message: any) => {
        const { data } = parseResponse(message.data)
        return Requester.success('1', { data: { result: data.p } })
      },
      onConnectChain: [
        {
          payload: `login|${defaultConfig.apiKey}`,
          filter: (message: any) => message.data.indexOf('post_login_success') !== -1,
        },
      ],
    }
  }
}
