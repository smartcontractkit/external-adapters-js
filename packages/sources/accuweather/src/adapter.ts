import { Builder } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
