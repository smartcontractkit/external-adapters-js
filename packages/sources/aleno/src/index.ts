import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
