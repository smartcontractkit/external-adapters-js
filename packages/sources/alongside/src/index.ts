import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
