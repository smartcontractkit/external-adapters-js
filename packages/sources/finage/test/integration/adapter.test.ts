import { Requester } from '@chainlink/ea-bootstrap'
import { assertError, assertSuccess } from '@chainlink/ea-test-helpers'
import { AdapterRequest } from '@chainlink/types'
import { makeExecute } from '../../src/adapter'

describe('execute', () => {
  const jobID = '1'
  const execute = makeExecute()
  process.env.API_KEY = process.env.API_KEY ?? 'test_api_key'

  describe('successful calls @integration', () => {
    const requests = [
      {
        name: 'id not supplied',
        testData: { data: { symbol: 'AAPL' } },
      },
      {
        name: 'symbol AAPL',
        testData: { id: jobID, data: { symbol: 'AAPL' } },
      },
      {
        name: 'symbol TSLA',
        testData: { id: jobID, data: { symbol: 'TSLA' } },
      },
      {
        name: 'from',
        testData: { id: jobID, data: { from: 'AAPL' } },
      },
      {
        name: 'endpoint stock',
        testData: { id: jobID, data: { base: 'TSLA', endpoint: 'stock' } },
      },
      {
        name: 'endpoint eod',
        testData: { id: jobID, data: { symbol: 'VXX', endpoint: 'eod' } },
      },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        const data = await execute(req.testData as AdapterRequest)
        assertSuccess({ expected: 200, actual: data.statusCode }, data, jobID)
        expect(data.result).toBeGreaterThan(0)
        expect(data.data.result).toBeGreaterThan(0)
      })
    })
  })

  describe('successful batch calls', () => {
    const requests = [
      {
        name: 'supports multiple symbols',
        testData: { id: jobID, data: { from: ['TSLA', 'AAPL'], endpoint: 'stock' } },
      },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        const data = await execute(req.testData as AdapterRequest)
        assertSuccess({ expected: 200, actual: data.statusCode }, data, jobID)
        expect(Object.keys(data.data.result).length).toBeGreaterThan(0)
      })
    })
  })

  describe('validation error', () => {
    const requests = [
      { name: 'empty body', testData: {} },
      { name: 'empty data', testData: { data: {} } },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        try {
          await execute(req.testData as AdapterRequest)
        } catch (error) {
          const errorResp = Requester.errored(jobID, error)
          assertError({ expected: 400, actual: errorResp.statusCode }, errorResp, jobID)
        }
      })
    })
  })

  describe('error calls @integration', () => {
    const requests = [
      {
        name: 'unknown symbol',
        testData: { id: jobID, data: { base: 'not_real' } },
      },
    ]

    requests.forEach((req) => {
      it(`${req.name}`, async () => {
        try {
          await execute(req.testData as AdapterRequest)
        } catch (error) {
          const errorResp = Requester.errored(jobID, error)
          assertError({ expected: 500, actual: errorResp.statusCode }, errorResp, jobID)
        }
      })
    })
  })
})
