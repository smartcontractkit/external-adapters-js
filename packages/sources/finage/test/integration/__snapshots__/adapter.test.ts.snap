// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute crypto api should return success 1`] = `
{
  "data": {
    "result": 50940.12,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 50940.12,
  "statusCode": 200,
}
`;

exports[`execute eod api should return success 1`] = `
{
  "data": {
    "result": 26.3,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 26.3,
  "statusCode": 200,
}
`;

exports[`execute forex api should return success 1`] = `
{
  "data": {
    "result": 1.3443049999999999,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 1.3443049999999999,
  "statusCode": 200,
}
`;

exports[`execute stock api should return success 1`] = `
{
  "data": {
    "result": 25.8,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 25.8,
  "statusCode": 200,
}
`;

exports[`execute stock api with invalid base should return failure 1`] = `
{
  "error": {
    "errorResponse": "Please check the symbol and try again.",
    "feedID": "{"data":{"endpoint":"stock","base":"non-existing"}}",
    "message": "Request failed with status code 400",
    "name": "AdapterError",
    "url": "https:/api.finage.co.uk/last/stock/NON-EXISTING",
  },
  "jobRunID": "1",
  "providerStatusCode": 400,
  "status": "errored",
  "statusCode": 200,
}
`;

exports[`execute commodities api should return success 1`] = `
{
  "data": {
    "result": 98.91,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 98.91,
  "statusCode": 200,
}
`;

exports[`execute commodities api with invalid base should return failure 1`] = `
{
  "error": {
    "errorResponse": "Please check the symbol and try again.",
    "feedID": "NONEXISTING/USD",
    "message": "Request failed with status code 400",
    "name": "AdapterError",
    "url": "https:/api.finage.co.uk/last/trade/forex/NONEXISTINGUSD",
  },
  "jobRunID": "1",
  "providerStatusCode": 400,
  "status": "errored",
  "statusCode": 200,
}
`;